File: prototype/utility.php
<?php

/***************************************************************************
    Returns character offset of the first mismatch between two strings
***************************************************************************/
function strdiff( $string1, $string2 )
{
    $length = min( strlen( $string1 ), strlen( $string2 ) );

    for ( $i = 0; $i < $length; $i++ )
    {
        if ( $string1{ $i } != $string2{ $i } )
        {
            break;
        }
    }

    return $i;
}

/***************************************************************************
    Returns gretest common divider of positive integers array
***************************************************************************/
function gcd()
{
    if ( ( func_num_args() == 1 ) && is_array( func_get_arg( 0 ) ) )
    {
        $argument = func_get_arg( 0 );
    }
    else
    {
        $argument = func_get_args();
    }

    foreach ( $argument as $a )
    {
        $number[] = intval( $a );
    }

    $result = max( min( $number ), 0 );
    $count  = count( $number );

    for ( $i = 0; ( $result > 1 ) && ( $i < $count ); $i++ )
    {
        $dividend = $number[ $i ];
        $reminder = $dividend % $result;

        while ( $reminder != 0 )
        {
            $dividend = $result;
            $result   = $reminder;
            $reminder = $dividend % $result;
        }
    }

    return $result;
}

/***************************************************************************
    Returns array of prime factors of an integer
***************************************************************************/
function primefact( $number )
{
    $result = array();
    $factor = 2;

    while ( ( $factor * $factor ) <= $number )
    {
        if ( ( $number % $factor ) == 0 )
        {
            $result[] = $factor;
            $number  /= $factor;
        }
        else
        {
            $factor++;
        }
    }

    $result[] = $number;

    return $result;
}

/***************************************************************************
    Returns array of all possible factor pairs of an integer
***************************************************************************/
function pairfact( $number )
{
    $prime  = primefact( $number );
    $primes = count( $prime );
    $setSize1 = 1;
    $setSize2 = $primes - $setSize1;
    $factor   = array( 1 => $number );

    if ( $primes == 2 )
    {
        $factor[ $prime[ 0 ] ] = $prime[ 1 ];
    }

    while ( $setSize1 <= $setSize2 )
    {
        for ( $i = 0; $i < ( $primes - ( $setSize1 - 1 ) ); $i++ )
        {
            $factor1 = 1;
            $factor2 = 1;

            for ( $j = 0; $j < $setSize1; $j++ )
            {
                $factor1 *= $prime[ $i + $j ];
            }

            for ( $k = 0; $k < $i; $k++ )
            {
                $factor2 *= $prime[ $k ];
            }

            for ( $k = $i + $setSize1; $k < $primes; $k++ )
            {
                $factor2 *= $prime[ $k ];
            }

            $factor[ min( $factor1, $factor2 ) ] = max( $factor1, $factor2 );
        }

        $setSize1++;
        $setSize2 = $primes - $setSize1;
    }

    foreach ( $factor as $factor1 => $factor2 )
    {
        $result[] = array( $factor1, $factor2 );
    }

    return $result;
}

/***************************************************************************
    Returns a cosine fade-out gain
***************************************************************************/
function cosfade( $number, $start, $end )
{
    if ( $number <= $start )
    {
        $result = 1.0;
    }
    else
    {
        if ( $number >= $end )
        {
            $result = 0.0;
        }
        else
        {
            $result = 0.5 * ( cos( pi() * ( $number - $start ) / ( $end - $start ) ) + 1.0 );
        }
    }

    return $result;
}

File: classifier/src/classifier/utility.cpp
/*******************************************************************************
*******************************************************************************/

#include <QList>
#include <QMap>
#include <QPair>
#include <QSet>
#include <QString>
#include <math.h>
#include "utility.h"

/*******************************************************************************
    Returns character offset of the first mismatch between two strings
*******************************************************************************/
int difference
(
    QString const & string1,
    QString const & string2
)
{
    int           result = 0;
    QChar const * data1  = string1.constData();
    QChar const * data2  = string2.constData();
    int           length = qMin( string1.length(), string2.length() );

    while ( ( result < length ) && ( *data1++ == *data2++ ) )
    {
        ++result;
    }

    return result;
}

/*******************************************************************************
    Returns array of prime factors of an integer
*******************************************************************************/
QIntList primeFactor
(
    int const number
)
{
    QIntList   result;
    int        factor   = 2;
    int        dividend = number;

    while ( ( factor * factor ) <= dividend )
    {
        if ( ( dividend % factor ) == 0 )
        {
            result.append( factor );

            dividend /= factor;
        }
        else
        {
            ++factor;
        }
    }

    result.append( dividend );

    return result;
}

/*******************************************************************************
    Returns array of all possible factor pairs of an integer
*******************************************************************************/
QIntPairList pairFactor
(
    int const number
)
{
    QMap< int, int >   factors;
    QIntList           primes   = primeFactor( number );
    int                setSize1 = 1;
    int                setSize2 = primes.count() - setSize1;

    factors[ 1 ] = number;

    if ( primes.count() == 2 )
    {
        factors[ primes[ 0 ] ] = primes[ 1 ];
    }

    while ( setSize1 <= setSize2 )
    {
        for ( int i = 0; i < ( primes.count() - ( setSize1 - 1 ) ); i++ )
        {
            int   factor1 = 1;
            int   factor2 = 1;

            for ( int j = 0; j < setSize1; j++ )
            {
                factor1 *= primes[ i + j ];
            }

            for ( int k = 0; k < i; k++ )
            {
                factor2 *= primes[ k ];
            }

            for ( int k = i + setSize1; k < primes.count(); k++ )
            {
                factor2 *= primes[ k ];
            }

            factors[ qMin( factor1, factor2 ) ] = qMax( factor1, factor2 );
        }

        setSize2 = primes.count() - ++setSize1;
    }

    QIntPairList result;

    foreach ( int factor, factors.keys() )
    {
        result.append( QIntPair( factor, factors[ factor ] ) );
    }

    return result;
}

/*******************************************************************************
    Returns a cosine fade-out gain
*******************************************************************************/
float cosfade
(
    int const value,
    int const fadeStart,
    int const fadeEnd
)
{
    static float const pi = 3.14159265358f;

    float result;

    if ( value <= fadeStart )
    {
        result = 1.0f;
    }
    else
    {
        if ( value >= fadeEnd )
        {
            result = 0.0f;
        }
        else
        {
            result = 0.5f * ( cos( pi * ( value - fadeStart ) / ( fadeEnd - fadeStart ) ) + 1.0f );
        }
    }

    return result;
}

/*******************************************************************************
    Returns gretest common divider of two positive integers
*******************************************************************************/
int gcd
(
    int const number1,
    int const number2
)
{
    int   result   = qMin( number1, number2 );
    int   dividend = qMax( number1, number2 );
    int   reminder = dividend % result;

    while ( reminder != 0 )
    {
        dividend = result;
        result   = reminder;
        reminder = dividend % result;
    }

    return result;
}

/*******************************************************************************
    Returns gretest common divider of list of positive integers
*******************************************************************************/
int gcd
(
    QIntList const & numbers
)
{
    int result = 0;

    if ( numbers.count() > 0 )
    {
        QIntListIterator iterator( numbers );

        result = iterator.next();

        while ( ( result > 1 ) && iterator.hasNext() )
        {
            result = gcd( result, iterator.next() );
        }
    }

    return result;
}

/*******************************************************************************
    Returns gretest common divider of set of positive integers
*******************************************************************************/
int gcd
(
    QIntSet const & numbers
)
{
    int result = 0;

    if ( numbers.count() > 0 )
    {
        QIntSetIterator iterator( numbers );

        result = iterator.next();

        while ( ( result > 1 ) && iterator.hasNext() )
        {
            result = gcd( result, iterator.next() );
        }
    }

    return result;
}

/******************************************************************************/
